README
URL::Check

The README  is  used  to  introduce the module and provide instructions on how to
install the module, any machine dependencies it may have (for example C compilers
and installed libraries) and any other information that should be provided before
the module is installed.

A  README  file  is required for CPAN modules since CPAN extracts the README file
from a module distribution so that people browsing the archive can use it to  get
an  idea  of  the  module's  uses.  It  is usually a good idea to provide version
information here so that people can decide whether fixes for the module are worth
downloading.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing,  you  can  find  documentation for this module with the perldoc
command.

    perldoc URL::Check

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=URL-Check

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/URL-Check

    CPAN Ratings
        http://cpanratings.perl.org/d/URL-Check

    Search CPAN
        http://search.cpan.org/dist/URL-Check/


LICENSE AND COPYRIGHT

Copyright (C) 2011 Alexandre Masselot

This program is free software; you can redistribute it and/or modify it under the
terms of either: the GNU General Public License as published by the Free Software
Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

Changes
Revision history for URL::Check

0.10   Wed Dec 16 10:50:00 2015 MANWAR
       - Updated Copyright year information.
       - Tidied up pod document in general.

0.09   Mon Nov 30 11:00:00 2015 MANWAR
       - Earlier fix didn't resolve the issue, so one more attempt to resolve it,
         using module Capture::Tiny.

0.08   Thu Nov 26 10:40:00 2015 MANWAR
       - Enable warning check in the unit test script t/02-test-to-console.t.
         Also changed slightly how the STDOUT is captured as some CPAN Tester
         reporting error on amd64-netbsd as below:
         http://www.cpantesters.org/cpan/report/8a4733de-9220-11e5-afef-e7dd1d056eec

0.07   Sat Nov 21 11:40:00 2015 MANWAR
       - Tidied up script/url-check script.

0.06   Sat Nov 21 11:20:00 2015 MANWAR
       - Added $AUTHORITY to the package URL::Check.
       - Added key 'provides' to the Makefile.PL script.
       - Added LICENSE to the MANIFEST file.
       - Removed 'Test::More' from prereqs list in the Makefile.PL script.

0.05   Fri Nov 20 12:55:00 2015 MANWAR
       - Copied 'script/url-check.pl' and created another called 'script/url-check'.
         Keeping the old one as it is just to be backward compatible.
       - On install, copy the new script 'url-check' to the user path. So that it can
         be run from anywhere.
       - Updated pod document with regard to the above changes.

0.04   Thu Nov 19 10:30:00 2015 MANWAR
       - Added LICENSE file.
       - Added key 'LICENSE' to the Makefile.PL
       - Added section "REPOSITORY" to the pod document.
       - Added section "MAINTAINER" to the pod document.
       - General tidied up pod document.

0.03   Wed Nov 18 10:25:00 2015 MANWAR
       - Tidied up Makefile.PL script.
       - Removed Build.PL script.
       - Tidied up README file.
       - Added github repository to the Makefile.PL file.

0.02   2011-10-02 ALEXMASS
       - Some more changes.

0.01   2011-10-02 ALEXMASS
       -  First version, released on an unsuspecting world.


lib/URL/Check.pm
package URL::Check;

$URL::Check::VERSION   = '0.10';
$URL::Check::AUTHORITY = 'cpan:ALEXMASS';

=head1 NAME

URL::Check - Check a list of URL and respond accordingly.

=head1 VERSION

Version 0.10

=cut

use 5.006;
use strict; use warnings;

use LWP::Simple qw/get/;
use Time::HiRes qw /gettimeofday/;

=head1 DESCRIPTION

This module is not aimed at being used directly but via the script C<url-check>.

    url-check --config=my-config.txt

If no C<--config argument> is set, the value is taken from environment variable C<URL_CHECK_CONFIG>.

More examples can be found in the C<t/resources/config> directory but consist
in default parameter (mailto etc., then each url to be tested can be followed
by dedicated test (time.delay, xpath etc.)

More info can be found L<here|http://alexandre-masselot.blogspot.com/2011/10/perl-url-checker.html>

=head1 METHODS

=head2 readConfig($file_name)

Read the config file. Default file name is taken from C<$URL_CHECK_FILE>

=cut

our %config;
our @report;

sub readConfig {
    my $configFile = shift || $ENV{URL_CHECK_CONFIG}
    || die "ERROR: No config file is passed or env URL_CHECK_CONFIG is set.\n";

    my $FD;
    open ($FD, "<$configFile")
        || die "ERROR: Cannot open config file [$configFile]: $!\n";

    p_clearConfig();

    # $currenturl will contain at first the default map for handling error,
    # then, when a url is encountered, it will contain the url's error handlers
    my $currentUrl;
    while (my $line = <$FD>) {
        $line=~s/^#.*//;		# remove comments
        $line=~s/\s*$//;		# end of line spaces
        next unless $line =~/\S/;	# skip emplty lines

        if ($line=~/^onerror/i) {
            p_addOnErrorLine($line, $currentUrl || $config{default});
            next;
        }

        if ($line=~/^check/i) {
            p_addCheckLine($line, $currentUrl);
            next;
        }

        if ($line=~/^(ftp|http|file):\/\//i) {
            $currentUrl = p_addUrl($line);
            next;
        }

        die "ERROR: Cannot parse line: $line\n";
    }

    close($FD);
}

=head2 run()

Run the configured tests, and store the result into the local @report

=cut

sub run {

    undef @report;
    foreach my $urlConfig (@{$config{urls}}) {
      push @report, p_runOneUrl($urlConfig);
    }
}

=head2 submitReport(%report)

Print on console or send by mail the error output

=cut

sub submitReport {
    my %report = @_;

    if ($config{default}{onError}{console}) {
        print "ERROR REPORT: $report{subject}\n$report{contents}\n";
    }

    if ($config{default}{onError}{mailto}) {
        require  Mail::Sendmail;
        my $hostname  = `hostname`;
        chomp $hostname;
        my %mail = (
            To      => join(',', @{$config{default}{onError}{mailto}}),
            Subject => "[url-check] $report{subject}",
            From    => "url.check\@$hostname",
            Message => "$report{contents}\n",
        );

        warn "sending error report by mail: $report{subject}\n";
        Mail::Sendmail::sendmail(%mail) or die $Mail::Sendmail::error
    }
}

=head2 errorReport()

Build a map (subject => ..., content=> ...) with the errors after the run
return () if no error were detected

=cut

sub errorReport {
    my @errors = grep {! $_->{success}} @report;

    unless (@errors) {
        return ();
    }

    (
     subject  => ''.scalar(@errors).' errors reported',
     contents => join("\n", (map {$_->{url}." : ".$_->{message}} @errors))
    );
}

#
#
# PRIVATE METHODS

sub p_addOnErrorLine {
    my ($line, $conf) = @_;

    die "ERROR: Cannot parse error line: $line"
        unless $line=~/^onerror\.(.+?)=(.+)/i;

    my ($errorCat, $params) = ($1, $2);

    if ($errorCat eq 'mailto') {
        my @tmp = split(/,/, $params);
        $conf->{onError}{mailto}=\@tmp;
        return;
    }

    if ($errorCat eq 'console') {
        $conf->{onError}{console}= $params =~ /\s*(y(es)?|t(rue)?|1)\s*$/i;
        return;
    }

    die "ERROR: Unknown onerror type [$errorCat]\n";
}

sub p_addCheckLine {
    my ($line, $conf) = @_;

    die "ERROR: Cannot parse error line: $line\n"
        unless $line=~/^check\.(.+?)=(.+)/i;

    my ($cat, $params) = ($1, $2);

    if ($cat eq 'contains') {
        push @{ $conf->{check}{contains}}, $params ;
        return;
    }

    $conf->{check}{$cat} = $params
}

sub p_addUrl {
    my ($line) = @_;

    my $h = {
        url   => $line,
        check => {}
    };

    push @{$config{urls}}, $h;
    return $h;
}

sub p_clearConfig {

    %config = (
        default => { onError=>{} },
        urls    => [],
    );
}

sub p_runOneUrl {
    my %urlConfig = %{$_[0]};

    my $url              = $urlConfig{url};
    my ($sec0, $micros0) = gettimeofday;
    my $content          = get($url);

    unless ($content) {
        return {
	    url     => $url,
	    success => 0,
	    message => 'cannot load content'
        }
    }

    my ($sec1, $micros1) = gettimeofday;
    my $dtime = int(($sec1-$sec0)*1000 + ($micros1-$micros0)/1000);

    if ((exists $urlConfig{check}{overtime}) &&  ($urlConfig{check}{overtime} < $dtime)) {
        return {
	    url     => $url,
	    success => 0,
	    message => "overtime > $urlConfig{check}{overtime} (${dtime}ms)",
        }
    }

    if (exists $urlConfig{check}{contains}) {
        my @searchFor = @{$urlConfig{check}{contains}};
        foreach (@searchFor) {
            if (index($content, $_) < 0) {
                return {
                    url     => $url,
                    success => 0,
                    message => "does not contains \"$_\"",
	       }
            }
        }
    }

    return {
        url     => $url,
        success => 1
    };
}

=head1 AUTHOR

Alexandre Masselot, C<< <alexandre.masselot at gmail.com> >>

Currently maintained by Mohammad S Anwar (MANWAR) C<< <mohammad.anwar @ yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/URL-Check>

=head1 BUGS

Please report any bugs or feature requests to C<bug-url-check at rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=URL-Check>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc URL::Check

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=URL-Check>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/URL-Check>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/URL-Check>

=item * Search CPAN

L<http://search.cpan.org/dist/URL-Check/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2015 Alexandre Masselot.

This program is free software; you can redistribute it and/or modify it under the
terms of either: the GNU General Public License as published by the Free Software
Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1; # End of URL::Check


Makefile.PL
#!/usr/bin/perl

use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'URL::Check',
    AUTHOR             => q{Alexandre Masselot <alexmass@cpan.org>},
    VERSION_FROM       => 'lib/URL/Check.pm',
    ABSTRACT_FROM      => 'lib/URL/Check.pm',
    LICENSE            => 'artistic_2',
    EXE_FILES          => [ 'script/url-check' ],
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES      => {
        'Test::More'    => 0,
        'Capture::Tiny' => '0.28',
    },
    PREREQ_PM            => {
        'Mail::Sendmail' => 0,
        'Time::HiRes'    => 0,
        'LWP::Simple'    => 0
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'URL-Check-*' },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        provides    => { 'URL::Check' => { file => 'lib/URL/Check.pm', version => '0.10' } },
        resources   => {
            repository => {
                type => 'git',
                url  => 'https://github.com/manwar/URL-Check.git',
                web  => 'https://github.com/manwar/URL-Check',
            },
        }})
     : ()
    ),
);
